pkg/resource/deploy/builtins.go

```diff
@@ -0,0 +1,128 @@
+}
+
+func (p *builtinProvider) Pkg() tokens.Package {
+	return "pulumi"
+}
+
+func (p *builtinProvider) label() string {
+	return "Builtins"
+}
+
+// CheckConfig validates the configuration for this resource provider.
+func (p *builtinProvider) CheckConfig(olds, news resource.PropertyMap) (resource.PropertyMap, []plugin.CheckFailure, error) {
+	return nil, nil, nil
+}
+
+// DiffConfig checks what impacts a hypothetical change to this provider's configuration will have on the provider.
+func (p *builtinProvider) DiffConfig(olds, news resource.PropertyMap) (plugin.DiffResult, error) {
+
+	return plugin.DiffResult{Changes: plugin.DiffNone}, nil
+}
+
+func (p *builtinProvider) Configure(props resource.PropertyMap) error {
+	return nil
+}
+
+func (p *builtinProvider) Check(urn resource.URN, olds, news resource.PropertyMap,
+	allowUnknowns bool) (resource.PropertyMap, []plugin.CheckFailure, error) {
+
+	return nil, nil, errors.Errorf("unrecognized resource type '%v'", urn.Type())
+}
+
+func (p *builtinProvider) Diff(urn resource.URN, id resource.ID, olds, news resource.PropertyMap,
+	allowUnknowns bool) (plugin.DiffResult, error) {
+
+	contract.Fail()
+	return plugin.DiffResult{}, errors.Errorf("the builtin provider has no resources")
+}
+
+func (p *builtinProvider) Create(urn resource.URN,
+	news resource.PropertyMap) (resource.ID, resource.PropertyMap, resource.Status, error) {
+
+	contract.Fail()
+	return "", nil, resource.StatusUnknown, errors.Errorf("the builtin provider has no resources")
+}
+
+func (p *builtinProvider) Update(urn resource.URN, id resource.ID, olds,
+	news resource.PropertyMap) (resource.PropertyMap, resource.Status, error) {
+
+	contract.Fail()
+	return nil, resource.StatusUnknown, errors.Errorf("the builtin provider has no resources")
+}
+
+func (p *builtinProvider) Delete(urn resource.URN, id resource.ID, props resource.PropertyMap) (resource.Status, error) {
+	contract.Fail()
+	return resource.StatusUnknown, errors.Errorf("the builtin provider has no resources")
+}
+
+func (p *builtinProvider) Read(urn resource.URN, id resource.ID,
+	props resource.PropertyMap) (resource.PropertyMap, resource.Status, error) {
+	return nil, resource.StatusUnknown, errors.Errorf("the builtin provider has no resources")
+}
+
+func (p *builtinProvider) Invoke(tok tokens.ModuleMember,
+	args resource.PropertyMap) (resource.PropertyMap, []plugin.CheckFailure, error) {
+
+	switch tok.Name() {
+	case "getStack":
+		if p.backendClient == nil {
+			return nil, nil, errors.New("no backend client is available")
+		}
+
+		name, ok := args["name"]
+		if !ok {
+			return nil, []plugin.CheckFailure{{Property: "name", Reason: "missing required property 'name'"}}, nil
+		}
+		if !name.IsString() {
+			return nil, []plugin.CheckFailure{{Property: "name", Reason: "'name' must be a string"}}, nil
+		}
+
+		result, err := p.backendClient.GetStackOutputs(p.context, name.StringValue())
+		if err != nil {
+			return nil, nil, err
+		}
+		return result, nil, nil
```

Comment:
I don't think we should return the outputs directly from `getStack`.  There may be additional metadata we want to expose about a stack over time.  I'd suggest we return a struct which includes the outputs as a member - probably the only member for now.

```typescript
pulumi.service.getStack("lukehoban/myapp-production").outputs["url"]
// or
pulumi.service.getStack("lukehoban/myapp-production").outputs.url
```

Reply:
I've been going back-and-forth about this:
- Do you have any examples of the other data we might want to expose and how it would be used?
- If we want to add additional data, do you think that this API is the right interface?

Reply:
Other data:
* Resources (effectively similar to `.get`, but without having them be serialized into this stacks checkpoint)
* Version (or generally, some value that changes when the stack changes to trigger off of)
* LastUpdateTime (maybe that is a substitute for above)
* I would not be surprised if we have additional things here in the future - like transitive stack dependencies if we start tracking that.

I can see two options for API:
1. Similar to above - `getStack` that returns an object which we can grow over time, but which today only exposes `outputs`.
2. Just `getStackOutputs` for now, and we can add `getStack` later or add `getStackResources` or similar.

(2) might be nice as it also helps with the `Promise` issue discussed below.